src/SDK/SmartDetectorsShared/Clients/AnalysisServicesFactory.cs

```diff
@@ -184,19 +185,49 @@ private async Task<IReadOnlyList<string>> GetResourceIdsForLogAnalyticsAsync(IRe
+                // The workspaces associated with each resource have to be retrieved.
                 var workspacesList = new List<ResourceIdentifier>();
-                foreach (string subscriptionId in resources.Select(resource => resource.SubscriptionId).Distinct(StringComparer.CurrentCultureIgnoreCase))
+
+                foreach (ResourceIdentifier resource in resources)
                 {
-                    // Try to get the workspaces list from the cache, and if it isn't there, use the Azure Resource Manager client
-                    IList<ResourceIdentifier> subscriptionWorkspaces;
-                    if (!this.subscriptionIdToWorkspaces.TryGetValue(subscriptionId, out subscriptionWorkspaces))
+                    // There is no general way to get only the specific workspace containing the telemetry for any resource.
+                    // A method that works for the Azure kubernetes Service is implemented here. If a resource is not AKS then all workspaces in it's subscription must be retrieved.
+                    if (resource.ResourceType == ResourceType.KubernetesService)
                     {
-                        subscriptionWorkspaces = await this.azureResourceManagerClient.GetAllResourcesInSubscriptionAsync(subscriptionId, new[] { ResourceType.LogAnalytics }, cancellationToken);
-                        this.subscriptionIdToWorkspaces[subscriptionId] = subscriptionWorkspaces;
+                        // Retrieve the specific workspace the target AKS cluster belongs to.
+                        ResourceIdentifier workspace;
+                        if (!this.aksClusterIdToWorkspaces.TryGetValue(resource.ToString(), out workspace))
+                        {
+                            // Try to get the workspaces from the cache, and if it isn't there, use the Azure Resource Manager client
+                            ResourceProperties querryResult = await this.azureResourceManagerClient.GetResourcePropertiesAsync(resource, cancellationToken);
+                            if (querryResult.Properties["addonProfiles"]["omsagent"]["enabled"].ToObject<bool>())
+                            {
+                                string idstring = querryResult.Properties["addonProfiles"]["omsagent"]["config"]["logAnalyticsWorkspaceResourceID"].ToString();
+                                workspace = ResourceIdentifier.CreateFromResourceId(idstring);
+                                this.aksClusterIdToWorkspaces[resource.ToString()] = workspace;
+                            }
+                            else
+                            {
+                                throw new TelemetryDataClientCreationException("Specified cluster does not have OMS agent onboarded");
+                            }
+                        }
+
+                        workspacesList.Add(workspace);
+                    }
```

Comment:
no objection for this code, but it is not needed because:
a. Right now we don't have any deployments of the Monitoring Appliance (all detectors run on the central environment), so it won't run
b. We plan to change this whole class to use resource-centric query API, so this whole method will be deleted.

You can either revert or leave as-is - up to you